### PR TITLE
Completed implementation and testing

### PR DESCRIPTION
**Pull request by bwssytems**
This is now working properly and I am using it in the ha-bridge. This should be just a straight merge to master. Let me know if you have any questions.

### CODE REVIEW

**Comment by mob41** [comment]
Hi, thanks for your pull. I will briefly check these days

**Comment by a1aw** [comment]
Hi thanks for the pull. I will check these days

I have removed the comment and put them into the reviews

File: pom.xml [hunk 1]
```diff
@@ -146,4 +145,5 @@
    		</plugins>
  	</build>
  	-->
+  <version>1.0.0</version>
```
Comment (a1aw) [inline pom.xml]: We need to follow the version format
x.x.x-SNAPSHOT/BETA/RELEASE as maven requires

File: src/main/java/com/github/mob41/blapi/BLDevice.java [hunk 2]
```diff
@@ -370,93 +367,66 @@ public static void printBytes(byte[] data) {
...
-
-        byte[] encData = subbytes(data, 0x38, data.length);
-
-        if (debug)
-            log.debug("encDataLen=" + encData.length);
+        log.debug("auth AuthPayload initial bytes to send: {}", DatatypeConverter.printHexBinary(sendPayload.getPayload().getData()));
 
-        if (encData.length % 16 != 0) {
-            log.warn("TODO: Incompatible decryption with non-16 multiple bytes. Forcing to have 1024 bytes");
+        DatagramPacket recvPack = sendCmdPkt(10000, 2048, sendPayload);
```
Comment (a1aw) [inline src/main/java/com/github/mob41/blapi/BLDevice.java:383]: I would like to know the usage of increasing the buffer size from 88 to 2048. Is it about the huge size of receiving IR/RF signals?

File: src/main/java/com/github/mob41/blapi/RM2Device.java [hunk 3]
```diff
@@ -111,7 +106,8 @@ public RM2Device(String host, Mac mac) throws IOException {
      */
     public boolean enterLearning() throws IOException {
         EnterLearnCmdPayload cmdPayload = new EnterLearnCmdPayload();
-        DatagramPacket packet = sendCmdPkt(10000, cmdPayload);
+        @SuppressWarnings("unused")
+		DatagramPacket packet = sendCmdPkt(10000, cmdPayload);
```
Comment (a1aw) [inline src/main/java/com/github/mob41/blapi/RM2Device.java]: TODO: Receive and get error_code

File: src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java [hunk 4]
```diff
@@ -72,95 +72,117 @@
      * @param payload
      *            The data to be sent
      */
-    public CmdPacket(Mac targetMac, int count, byte[] id, byte[] iv, byte[] key, CmdPayload cmdPayload) {
+    public CmdPacket(Mac targetMac, int count, byte[] id, byte[] iv, byte[] key, AES theAes, CmdPayload cmdPayload) {
```
Comment (a1aw) [inline src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java]: TODO: Remove unnecessary variables "iv, key" and rename "theAes" to "aesInstance"

File: src/main/java/com/github/mob41/blapi/pkt/auth/AuthPayload.java [hunk 5]
```diff
@@ -34,6 +34,35 @@
...
+        data[0x30] = (byte) 'T';
+        data[0x31] = (byte) 'e';
+        data[0x32] = (byte) 's';
+        data[0x33] = (byte) 't';
+        data[0x34] = (byte) ' ';
+        data[0x35] = (byte) ' ';
+        data[0x36] = (byte) '1';
+    }
+
     public AuthPayload(byte[] devId, byte[] devName) {
```
Comment (a1aw) [inline src/main/java/com/github/mob41/blapi/pkt/auth/AuthPayload.java]: TODO: Below it is for custom devId and devName. Uncomment the code and remove the existing code.

**Review by a1aw — CHANGES_REQUESTED** [review]
Please check the comments to the code.

File: pom.xml (hunk 1 above)
Comment (bwssytems) [reply pom.xml]: That would be fine. I just did that for my own purposes of saying it is the first working version.

File: src/main/java/com/github/mob41/blapi/BLDevice.java (hunk 2 above)
Comment (bwssytems) [reply src/main/java/com/github/mob41/blapi/BLDevice.java:383]: It is about having enough room to capture data on the receive. It also needs to be a multiple of 16. The actual payload is cut down after it is received to the first null after any data.

File: src/main/java/com/github/mob41/blapi/RM2Device.java (hunk 3 above)
Comment (bwssytems) [reply src/main/java/com/github/mob41/blapi/RM2Device.java]: I haven't done anything with learning yet. I should try and test it now that I have an RM Mini.

File: src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java (hunk 4 above)
Comment (bwssytems) [reply src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java]: Per instance, we should only have one AES object that is referred too. IT only needs to be created on initial, then once again after the auth is called.

File: src/main/java/com/github/mob41/blapi/pkt/auth/AuthPayload.java (hunk 5 above)
Comment (bwssytems) [reply src/main/java/com/github/mob41/blapi/pkt/auth/AuthPayload.java]: It looks as this is not needed. When comparing to the pyhton impl, it is not used, just the 'Test 1'.

File: src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java (hunk 4 above)
Comment (a1aw) [reply src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java]: I mean the iv and key is not used here.

Comment (bwssytems) [reply src/main/java/com/github/mob41/blapi/pkt/CmdPacket.java]: Ahh yes, that would be true. Makes Sense to streamline.

**Comment by a1aw** [comment]
I think I can rebase and merge right now? The TODOs can be followed up later.

**Comment by bwssytems** [comment]
I can do those items on my branch being merged. I believe that is the preferred way

**Comment by bwssytems** [comment]
Made changes, take a look

**Pull request merged**